File: frogment/src/main/java/ch/grze/frogment/frogment/FrogmentData.java
package ch.grze.frogment.frogment;

import android.os.Parcel;
import android.os.Parcelable;

import ch.grze.frogment.State;

public class FrogmentData implements Parcelable {
    private final Class<? extends FrogmentInterface> clazz;
    private final State state;
    private final String tag;

    private FrogmentData(Class<? extends FrogmentInterface> clazz, State state, String tag) {
        this.clazz = clazz;
        this.state = state;
        this.tag = tag == null ? clazz.getCanonicalName() : tag;
    }

    public static FrogmentData forClass(Class<? extends FrogmentInterface> clazz) {
        return new Builder(clazz).build();
    }

    public Class<? extends FrogmentInterface> getClazz() {
        return clazz;
    }

    public State getState() {
        return state;
    }

    public String getTag() {
        return tag;
    }

    public static class Builder {
        private final Class<? extends FrogmentInterface> clazz;
        private State state = null;
        private String tag = null;

        public Builder(Class<? extends FrogmentInterface> clazz) {
            this.clazz = clazz;
        }

        public Builder state(State state) {
            this.state = state;
            return this;
        }

        public Builder tag(String tag) {
            this.tag = tag;
            return this;
        }

        public FrogmentData build() {
            return new FrogmentData(clazz, state, tag);
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeSerializable(this.clazz);
        dest.writeParcelable(this.state, flags);
        dest.writeString(this.tag);
    }

    protected FrogmentData(Parcel in) {
        this.clazz = (Class<? extends FrogmentInterface>) in.readSerializable();
        this.state = in.readParcelable(State.class.getClassLoader());
        this.tag = in.readString();
    }

    public static final Creator<FrogmentData> CREATOR = new Creator<FrogmentData>() {
        @Override
        public FrogmentData createFromParcel(Parcel source) {
            return new FrogmentData(source);
        }

        @Override
        public FrogmentData[] newArray(int size) {
            return new FrogmentData[size];
        }
    };
}

File: frogment/src/main/java/ch/grze/frogment/core/navigation/FrogmentData.kt
package ch.grze.frogment.core.navigation

import android.os.Parcel
import android.os.Parcelable
import ch.grze.frogment.State
import ch.grze.frogment.frogment.FrogmentInterface

class FrogmentData
@JvmOverloads constructor(
        val clazz: Class<out FrogmentInterface>,
        val state: State? = null,
        val tag: String? = clazz.canonicalName
) : Parcelable {
    private constructor (builder: Builder) : this(builder.clazz, builder.state, builder.tag)

    class Builder(val clazz: Class<out FrogmentInterface>) {
        internal var state: State? = null
        internal var tag: String = clazz.canonicalName

        fun state(state: State): Builder {
            this.state = state
            return this
        }

        fun tag(tag: String): Builder {
            this.tag = tag
            return this
        }

        fun build(): FrogmentData = FrogmentData(this)
    }

    constructor(source: Parcel) : this(source.readSerializable() as Class<out FrogmentInterface>, source.readParcelable<State>(State::class.java.classLoader), source.readString())

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeSerializable(clazz)
        writeParcelable(state, 0)
        writeString(tag)
    }

    companion object {
        @JvmStatic
        fun forClass(clazz: Class<out FrogmentInterface>): FrogmentData = FrogmentData(clazz)

        @JvmField
        val CREATOR: Parcelable.Creator<FrogmentData> = object : Parcelable.Creator<FrogmentData> {
            override fun createFromParcel(source: Parcel): FrogmentData = FrogmentData(source)
            override fun newArray(size: Int): Array<FrogmentData?> = arrayOfNulls(size)
        }
    }
}